Allow TMC baud rate override

sjasonsmith opened the pull request:
### Description

Allow overriding TMC baud rate by defining TMC_BAUD_RATE. This is prepatory to using the SoftwareSerial provided by the STM32 framework. Unlike the 1768 version of  SoftwareSerial, the STM32 implementation of SoftwareSerial does not force the baud rate to 19200.

Set this to 19200 for the BTT SKR Pro and BTT002, which are the only two boards currently using the STM32 HAL which define software serial pin numbers.

### Benefits

Allows using slower baud rates with STM32 and SoftwareSerial, to reduce the chance of bit timing errors.

### Related Issues

This is related to PR #15655.

sjasonsmith committed:
Allow overriding TMC baud rate

sjasonsmith committed:
Set SKR Pro and BTT002 TMC baud rates to 19200

sjasonsmith commented:
@LinoBarreca, if this is merged it will eliminate the need to force the baud rate in the platformio.ini file.

I chose not to add this to Configuration_adv.h, since this is more board-specific than something many users would need to customize.

gloomyandy commented:
It would be good to also set this for all of the LPC176x boards (I'm pretty sure they all use software serial) and we can then remove the hack to force it in the framework.

boelle commented:
do we have links to issues that this PR might solve?

sjasonsmith commented:
> do we have links to issues that this PR might solve?

This does not solve existing issues, it is a mitigation for issues that will appear when we switch to the Arduino_Core_STM32 version of SoftwareSerial. You will see that when upgrading to the STM32 platform 5.7.

The current SoftwareSerial implementations from FYSETC and BTT for STM32 lock the baud rate to 19200 and 14400, respectively. The upcoming Arduino_Core_STM32 SoftwareSerial does not do this for us.

SoftwareSerial on STM32 (and 176x) cannot run reliably at our 115200 baud, so we need to override that in one way or another. It can be done through preprocessor definitions in the platformio.ini, but it seems more correct to pass the correct baud rate to TMCStepper from our code.

sjasonsmith commented:
> It would be good to also set this for all of the LPC176x boards (I'm pretty sure they all use software serial) and we can then remove the hack to force it in the framework.

@gloomyandy, I have been hearing your name a lot as I have been poking at SoftwareSerial! You might be interested in this PR I have to the STM32 version of SoftwareSerial, since I hear this is all code derived from your original implementation. The STM32 broke half-duplex behavior, and I am basically reverting it to more closely resemble the LPC176x implementation.
https://github.com/stm32duino/Arduino_Core_STM32/pull/749

I can add the LPC176x boards to the PR, but will you really be able to remove the forced baud rate from the framework? I have no idea how widely that is used outside of Marlin.

LinoBarreca commented:
I think we will not need this PR if they merge the priority change on the softwareserial.

sjasonsmith commented:
> I think we will not need this PR if they merge the priority change on the softwareserial.

Other versions of SoftwareSerial force a low baud rate, even though they are setting preempt priority to 0. I am quite sure that I tested this with the SKR Pro as well. At 115200 I was able to communicate with TMC, but data was very often corrupt.

Software timing asynchronous serial signals is just is not very reliable. One bit is less than nine microseconds, and error accumulates through the entire byte (~30 interrupts with 3x oversampling). Small amounts of jitter easy add up to bad data.

I have ideas that could improve the SoftwareSerial algorithm to be a little more robust, but don't have current plans to work on it. It will never compare to an actual hardware UART.

LinoBarreca commented:
@sjasonsmith the proper fix is easy as long as the serial ISR is short and doesn't get interrupted by other ISRs. actually, from your oscilloscope tests, it's very, very short (in time) with the proper priority it could even handle 500000 baud..so...why do we even care about it?

sjasonsmith commented:
It is all about jitter. As long as anything in Marlin can disable interrupts, it is safer to use a slower speed.

Even if this is NOT 100% necessary, the change to the actual TMC code is trivial. I think it is a good change, even we didn't need it right now!

The changes to the pins files can be reverted later if reliability is increased in other ways.

LinoBarreca commented:
I don't know if our CPU has this but on some platforms you can suspend interrupts based on priority like "don't suspend anything below level X"

gloomyandy commented:
@sjasonsmith Yes I wrote the software serial code for the LPC176x (based on a suggestion from @p3p) I saw your PR and it makes sense to me. I think that removing the baud rate override would be a good thing as it basically subverts the serial API. It was one of those things that I wanted to do, but at the time I think it needed changes to the framework, the TMC library and Marlin. I was pretty busy just getting things tested and working and never got back to it.  As far as I know the main user of the LPC176x framework (and especially the software serial code) is Marlin, so I doubt if removing the override will cause too many problems.

I'd always opt to run at a lower speed with this sort of sample based system as it just makes things more robust. The comms to the TMC devices do not need to run at high speed, so no need to add the extra stress.

LinoBarreca commented:
> 
> I'd always opt to run at a lower speed with this sort of sample based system as it just makes things more robust. The comms to the TMC devices do not need to run at high speed, so no need to add the extra stress.

what kind of communication is done with the driver? I mean....I don't know how those work...I thought they were like "configurable parameters drivers"...a sort of "open serial, write or read config, close serial"....am I wrong or do they need serial for more than that?

p3p commented:
> The comms to the TMC devices do not need to run at high speed, so no need to add the extra stress.

I agree completely that SoftwareSerial is not robust enough for higher baud rates and there's nothing we can do about that but the TMC `MONITOR_DRIVER_STATUS` feature actively communicates with the drivers, and there have been issues with sever stuttering at low baud rate.

@LinoBarreca Only `MONITOR_DRIVER_STATUS` (which really shouldnt be turned on unless debuging) has active communication I think, otherwise its just 1 shot config.

LinoBarreca commented:
so why don't we just handle this differently?
if it's one shot only let's do this shot when no other isrs are running.

LinoBarreca commented:
and anyway..I guess that the configuration is written on some registers...a few bytes anyway.... making it constant and slow speed makes sense...

gloomyandy commented:
Just to be clear, for sampled software serial (like that used here) slow is probably better, but for the software serial used on the Arduino (which uses interrupt on change and then spins in a loop inside the interrupt handler reading the bits) then faster baud rates are probably better! So simply setting everything to use a slow baud rate is probably not an option.

LinoBarreca commented:
@gloomyandy do you have any clue why they choose to implement it as sampled in STM32? As far as I know we could have the same logic as in arduino (which is anyway a better logical use of the "interrupt") there must be a reason why they implemented it this way, no?

gloomyandy commented:
@LinoBarreca I suspect that they just hacked the code that already existed for the LPC176x (which they already used on the SKR V1.3 board).

Are all of the pins on the STM chips capable of doing interrupt on rising/falling edge (they are not on the LPC176x)? If not then they may not have had the pins that can generate interrupts available.

Have you looked at the Arduino code? I'm not sure doing things that way is a good idea, it runs for relatively long periods (time to send/receive 10 bits) with interrupts disabled. This means that you can't use it while printing (so no MONITOR_DRIVER_STATUS).

LinoBarreca commented:
@gloomyandy no. I didn't read arduino's code and I knew we were able to have interrupts on rising and falling. I am kinda ignorant about serial communication but I might have an idea to make it work without loops in the isr.
Tell me if you find problems in the logic, okay?
So, the idea: instead of polling periodically (according to baud rate) the pin to see its level 1/0 we could attach two interrupts on rising and falling. Inside each handler we take the ticks when it happened so, instead of saying "polled 3 times, found 3 times 1" we say "we turned to 0, last time that the pin was at 1 was at millis x it stayed at 1 for Y millis. At this baud rate it means we received 3 times 1"
Doing like this the isr for both rising and falling is short and it reconstructs the message by just looking at the elapsed time since last change....

sjasonsmith committed:
Update 176x and STM32 boards using SW Serial pins to define a reduced baud rate.
These currently may force a reduced baud rate in their SoftwareSerial libraries,
but this prepares them for this to be removed.

sjasonsmith commented:
I have updated this to include all LPC 176x and STM32 boards which use software serial. Currently the SoftwareSerial provided by LPC176x, BTT, and FYSETC all force a reduced baud rate, but this prepares them to work with a non-forcing version of SoftwareSerial.

gloomyandy commented:
@LinoBarreca before you do anything I think you need to spend some time studying the datasheets for the various STM32 devices. It seems that (some of them at least) connect the various GPIO ports to interrupt lines in a multiplexed fashion (which will complicate things), you also need to consider how using these interrupts may impact other Marlin features (like end stop interrupts), and how it will interact with any other parts of the the existing STM32 framework. I would also check carefully that all of the GPIO lines on all of the supported devices are actually capable of generating an interrupt (that is not the case on the LPC176x family). For details of how some devices multiplex things see: https://stm32f4-discovery.net/2014/08/stm32f4-external-interrupts-tutorial/

As to your proposed scheme I'm sure it can be made to work, but you will also need to add some logic to handle the timeout situation (just in case you get noise on the line or you start to listen mid way through a bit sequence). You will also need to sort out how you will transmit bits, ideally without running with interrupts disabled.

LinoBarreca commented:
@andy I might be wrong but most things in marlin don't use interrupts (on pins, I mean). For example the LCD buttons and the endstops, if I remember correctly, are polled by the temperature ISR which runs at 1khz (timer interrupt)

LinoBarreca commented:
> 
> As to your proposed scheme I'm sure it can be made to work, but you will also need to add some logic to handle the timeout situation (just in case you get noise on the line or you start to listen mid way through a bit sequence). You will also need to sort out how you will transmit bits, ideally without running with interrupts disabled.

I am no expert because I never handled serial at such low level but as regards to the "timeout" we could just monitor the serial start bit... and if the "bits received" aren't equal to bit length everything is discarded.
The running without interrupts disabled is actually possible. it means that the "serial manager" (the component which will handle ISRs high/low outputs and the send/receive for every "send" will expect an echo (because the send will trigger the ISRs as if they were received chars) so for every byte sent you expect one equal back.
I think it could be made to work. but as you said maybe not every STM32 chip has this rising/falling IRS...and since they had to make one class working on everything they went that poll-way

gloomyandy commented:
@LinoBarreca Marlin allows the use of interrupts for endstop detection (it is optional, but arguably more accurate), so as on the Arduino the use of software serial may interfere with that. 

As to running without disabling interrupts I was referring to the simple way that the Arduino code generates the serial bitstream. If you try to use the same mechanism with interrupts enabled you will hit issues with jitter, which will probably have to be addressed. 

For the timeout issue, there are ,many ways it can be done, but for a complete solution you will need to deal with it. 

The thing is all of this (multiplexed interrupt lines, timeouts etc, the many different STM32 chips that you need to deal with) adds to the complexity of what at a first glance seems like a simpler solution. As I said I'm sure it can be made to work (so long as you confirm that the chips actually support the interrupt on change on the required pins), but the oversampling solution has the advantage of only needing a timer and working with pretty much any gpio pin and possibly has less interaction with other features. 

For now I think the key thing is get some solution that works reasonably well with all of the various STM variants.
